src/main/java/uk/gov/hmcts/reform/idam/web/AppController.java

```diff
@@ -254,25 +257,51 @@ public RedirectView logout(final Map<String, Object> model) {
      * @should redirect to passwordReset view
      */
     @GetMapping(value = "/passwordReset")
-    public String getPasswordReset(@RequestParam("token") String token, @RequestParam("code") String code) {
-        return this.passwordReset(token, code);
+    public String getPasswordReset(@RequestParam("token") String token, @RequestParam("code") String code, Model model) {
+        return this.passwordReset(token, code, model);
     }
 
     /**
      * @should redirect to reset password page if token is valid
      * @should redirect to token expired page if token is invalid
+     * @should redirect to token expired page if token is expired
      */
     @PostMapping(value = "/passwordReset")
-    public String passwordReset(@RequestParam("token") String token, @RequestParam("code") String code) {
-        String nextPage = RESETPASSWORD_VIEW;
+    public String passwordReset(@RequestParam("token") String token, @RequestParam("code") String code, Model model) {
+        String nextPage = EXPIRED_PASSWORD_RESET_LINK_VIEW;
         try {
             spiService.validateResetPasswordToken(token, code);
-        } catch (Exception e) {
-            nextPage = EXPIRED_PASSWORD_RESET_LINK_VIEW;
+            nextPage = RESETPASSWORD_VIEW;
+        } catch (HttpClientErrorException e) {
+            if (e.getStatusCode().value() == 404)
+            {
+                model.addAttribute("forgotPasswordLink", buildUrlFromTheBody(e.getResponseBodyAsString()));
+            }
+        } catch (IOException e) {
+            //return expired password reset link view
         }
         return nextPage;
     }
```

Comment:
not sure about this method, maybe the nextPage made it a bit harder to read
can we maybe drop the nextPage variable and just use return instead?

also not sure about using IOException instead of Exception? there's isnt a catch all, should that be handled by the controller advice in that case?